### DISCUSSION TITLE
About `:`

### DISCUSSION THREAD

**Discussion by hchiba1** [pg-format/pg-formatter#41] [General]
`:` is the core of the grammar.
So it is worth discussing.

**Comment by hchiba1** [comment]
I think the simplest solution is to disallow `:` in unquoted strings.

Using `:` in unquoted strings can be too confusing.

**Comment by hchiba1** [comment]
One reason to allow `:` is for URIs.

**Reply by hchiba1** [reply]
Allowing `:` in node IDs might be useful when the IDs are URIs.

**Comment by hchiba1** [comment]
When handwriting in the Blitzboard sandbox, `:` is allowed in property values.
However, when importing an external PG file into the Blitzboard sandbox, unquoted property values that include `:` will be quoted.

**Reply by hchiba1** [reply]
Is that right? @matsusho070

**Reply by matsusho070** [reply]
Right. It's just to avoid confusion of users.

**Reply by hchiba1** [reply]
Thanks!
What was the initial motivation to allow `:` in unquoted values?

**Reply by matsusho070** [reply]
As far as I remember, there is no particular reason to allow `:`. The sandbox allows `:` simply because PG parser allows it.

**Reply by hchiba1** [reply]
I think there was some background of this commit
https://github.com/blitzboard/blitzboard/commit/fc6e0ea

**Reply by matsusho070** [reply]
Sorry, I was mistaken. I don't remember the exact reason why we allowed `:` in the past, but it was probably to make it easier to write URLs. However, thinking about it now, I feel that `:` should be quoted.

**Comment by hchiba1** [comment]
I'm testing the following behavior:
* allowing `:` in unquoted IDs (`:a` `a:` `a:b` `:` are valid IDs)
* allowing `:` in unquoted property values (although I would not like to recommend using `:` in unquoted property values for less readability)
* disallowing `:` in unquoted property keys

**Reply by hchiba1** [reply]
In this setting, `a b:c:d` is valid and parsed as `{"type": "node", "id": "a", "labels": [], "properties": {"b": ["c:d"]}}`.

**Reply by hchiba1** [reply]
Is it possible to say (in spec or docs) that `a b:c:d` is valid, but `a b:"c:d"` is recommended for readability? It would be more like a coding convention.

**Reply by hchiba1** [reply]
I suppose allowing `:a` `a:` `a:b` `:` as IDs is not harmful.

**Reply by nichtich** [reply]
`a:b:c` could be read as `a: b:c` or as `a:b: c`. The spec must tell which to choose. Current grammar subset in pgraph implements the second variant, but this can be changed:

~~~pegjs
  name:( $PlainStart $( ( !":" PlainChar )* ":" )+ ) {
      return name.join("").slice(0,-1)
    }
~~~

Thinking about the coice I assume that a key with colon is less frequent than a value with colon, e.g.

~~~
github :website url:https://github.com/    # condense form
github :website url: https://github.com/   # more readable
github :website url:"https://github.com/"  # recommended because value contains colon
github :website url: "https://github.com/" # also recommended

~~~

**Reply by hchiba1** [reply]
Those are good examples. It seems reasonable that
```
github :website url:https://github.com/
```
is parsed as:
```
{"type": "node", "id": "github", "labels": ["website"], "properties": {"url": ["https://github.com/"]}}
```

**Reply by nichtich** [reply]
I've changed parsing and serializing rules in pgraphs 0.4.2 as discussed. Note that unquoted property keys can still include colon when the colon seperating key and value is preceded or followed by space:

~~~
a b:c:d     # a "b" : "c:d"
a b:c: d    # a "b:c" : "d"
a b:c :d    # a "b:c" : "d"
~~~
